google-cloud-storage/src/test/java/com/google/cloud/storage/DefaultBufferedWritableByteChannelTest.java

```diff
@@ -0,0 +1,549 @@
+   *   <li>data size is evenly divisible by buffer size and write size
+   *   <li>buffer size is larger than write size
+   *   <li>buffer size is not evenly divisible by write size
+   * </ul>
+   */
+  @Example
+  void scenario_c() throws IOException {
+    bufferingEagerlyFlushesWhenFull(WriteOps.of(105, 15, 7));
+  }
+
+  /**
+   * Scenario D:
+   *
+   * <ul>
+   *   <li>write and buffer size are smaller than data
+   *   <li>data size is not evenly divisible by either write size nor buffer size
+   *   <li>buffer size is smaller than write size
+   *   <li>write size is not evenly divisible by buffer size
+   * </ul>
+   */
+  @Example
+  void scenario_d() throws IOException {
+    bufferingEagerlyFlushesWhenFull(WriteOps.of(61, 3, 16));
+  }
+
+  /**
+   * Scenario E:
+   *
+   * <p>Some flushes are only partially consumed. Ensure we proceed with consuming the buffer
+   * provided to {@code write}
+   *
+   * <pre>
+   *           0                        27
+   * data:    |--------------------------|
+   *               5       14 17        27
+   * writes:  |----|--------|--|---------|
+   *                   10
+   * flush 1: |---------|
+   *                     12
+   * flush 2:   |---------|
+   *                               22
+   * flush 3:             |---------|
+   *                                    27
+   * flush 4:                     |------|
+   * </pre>
+   */
+  @Example
+  void partialFlushOfEnqueuedBytesFlushesMultipleTimes() throws IOException {
+    ByteBuffer buffer = ByteBuffer.allocate(10);
+
+    ByteBuffer data1 = DataGenerator.base64Characters().genByteBuffer(5);
+    ByteBuffer data2 = DataGenerator.base64Characters().genByteBuffer(9);
+    ByteBuffer data3 = DataGenerator.base64Characters().genByteBuffer(3);
+    ByteBuffer data4 = DataGenerator.base64Characters().genByteBuffer(10);
+
+    ImmutableList<ByteBuffer> buffers = ImmutableList.of(data1, data2, data3, data4);
+
+    int allDataSize = buffers.stream().mapToInt(ByteBuffer::remaining).sum();
+    byte[] allData =
+        buffers.stream().reduce(ByteBuffer.allocate(allDataSize), ByteBuffer::put).array();
+    buffers.forEach(b -> b.position(0));
+
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        CountingWritableByteChannelAdapter adapter =
+            new CountingWritableByteChannelAdapter(Channels.newChannel(baos));
+        BufferedWritableByteChannel c = new DefaultBufferedWritableByteChannel(buffer, adapter)) {
+
+      c.write(data1);
+      adapter.nextWriteMaxConsumptionLimit = 2L;
+      c.write(data2);
+
+      c.write(data3);
+      adapter.nextWriteMaxConsumptionLimit = 7L;
+      c.write(data4);
+
+      c.close();
+      assertThrows(ClosedChannelException.class, () -> c.write(null));
+
+      assertWithMessage("Unexpected total flushed length")
+          .that(adapter.writeEndPoints)
+          .isEqualTo(ImmutableList.of(2L, 12L, 19L, 27L));
+      assertThat(baos.toByteArray()).isEqualTo(allData);
```

Comment:
I thought this was a partial flush, shouldn't you assert that it's equal to a subset of allData?

Reply:
There are individual partial flushes, but we still end up with all the data. For a dataSize of 27 and buffer size of 10, we should be able to consume everything in three flushes, however due to two of the flushes being partial we end up need 4 flushes, with some overlap of ranges.